crates/walrus-service/src/client/daemon/routes.rs

```diff
@@ -373,6 +410,236 @@ pub(super) fn daemon_cors_layer() -> CorsLayer {
         .allow_headers(Any)
 }
 
+/// Retrieve a blob from quilt by its QuiltPatchId.
+///
+/// Takes a quilt patch ID and returns the corresponding blob from the quilt.
+/// The blob content is returned as raw bytes in the response body, while metadata
+/// such as the patch identifier and tags are returned in response headers.
+///
+/// # Example
+/// ```bash
+/// curl -X GET "http://localhost:31415/v1/blobs/by-quilt-patch-id/\
+/// DJHLsgUoKQKEPcw3uehNQwuJjMu5a2sRdn8r-f7iWSAAC8Pw"
+/// ```
+///
+/// Response:
+/// ```text
+/// HTTP/1.1 200 OK
+/// Content-Type: application/octet-stream
+/// X-Quilt-Patch-Identifier: my-file.txt
+/// ETag: "DJHLsgUoKQKEPcw3uehNQwuJjMu5a2sRdn8r-f7iWSAAC8Pw"
+///
+/// [raw blob bytes]
+/// ```
+#[tracing::instrument(level = Level::ERROR, skip_all)]
+#[utoipa::path(
+    get,
+    path = QUILT_PATCH_BY_ID_GET_ENDPOINT,
+    params(
+        (
+            "quilt_patch_id" = String, Path,
+            description = "The QuiltPatchId encoded as URL-safe base64",
+            example = "DJHLsgUoKQKEPcw3uehNQwuJjMu5a2sRdn8r-f7iWSAAC8Pw"
+        )
+    ),
+    responses(
+        (
+            status = 200,
+            description = "The blob was retrieved successfully. Returns the raw blob bytes, \
+                        the identifier and other attributes are returned as headers.",
+            body = [u8]
+        ),
+        GetBlobError,
+    ),
+    summary = "Get blob from quilt",
+    description = "Retrieve a specific blob from a quilt using its QuiltPatchId. Returns the \
+                raw blob bytes, the identifier and other attributes are returned as headers.",
+)]
+pub(super) async fn get_blob_by_quilt_patch_id<T: WalrusReadClient>(
+    request_headers: HeaderMap,
+    State((client, response_header_config)): State<(Arc<T>, Arc<AggregatorResponseHeaderConfig>)>,
+    Path(QuiltPatchIdString(quilt_patch_id)): Path<QuiltPatchIdString>,
+) -> Response {
+    let quilt_patch_id_str = quilt_patch_id.to_string();
+    tracing::debug!("starting to read quilt patch: {}", quilt_patch_id_str);
+
+    match client.get_blobs_by_quilt_patch_ids(&[quilt_patch_id]).await {
+        Ok(mut blobs) => {
+            if let Some(blob) = blobs.pop() {
+                build_quilt_patch_response(
+                    blob,
+                    &request_headers,
+                    &quilt_patch_id_str,
+                    &response_header_config,
+                )
+            } else {
+                tracing::info!("no blob returned for the requested quilt blob ID");
```

Comment:
Nit: maybe also log the patch id? Should this be in DEBUG level as well? Same below.